### PR TITLE
Adding pure CTC or attention mode

### PR DESCRIPTION
**Pull request by sw005320**
Switch to use pure CTC or attention mode upon requests at #133. Now it is switched by specifying `--mtlalpha 0.0` (pure attention) or `--mtlalpha 1.0` (pure CTC).

### CODE REVIEW

**Comment by sw005320** [comment]
Also changing warpctc as default in chainer by following the discussion in #115

**Comment by sw005320** [comment]
@arendu and @m-wiesner, now I'm finishing only CTC or attention mode. Please let me know if you have any comments.

File: src/nets/e2e_asr_attctc.py [hunk 1]
```diff
@@ -75,7 +75,12 @@ def __call__(self, x):
...
         alpha = self.mtlalpha
-        self.loss = alpha * loss_ctc + (1 - alpha) * loss_att
+        if alpha == 0:
+            self.loss = loss_att
+        elif alpha == 1:
+            self.loss = loss_ctc
+        else:
+            self.loss = alpha * loss_ctc + (1 - alpha) * loss_att
 
         if self.loss.data < CTC_LOSS_THRESHOLD and not math.isnan(self.loss.data):
```
Comment (kan-bayashi) [inline src/nets/e2e_asr_attctc.py]: Is it necessary to report ctc or att loss in pure ctc or att mode?

File: egs/chime5/asr1/run.sh [hunk 2]
```diff
@@ -27,7 +27,6 @@ elayers=6
 eunits=320
 eprojs=320
 subsample=1_2_2_1_1 # skip every n frame from input to nth layers
-# loss related
 ctctype=warpctc
```
Comment (kan-bayashi) [inline egs/chime5/asr1/run.sh]: may be removed?

File: src/asr/asr_chainer.py [hunk 3]
```diff
@@ -344,6 +336,16 @@ def train(args):
...
 
+    # specify attention, CTC, hybrid mode
+    if args.mtlalpha == 1.0:
+        mtl_mode = 'ctc'
+        logging.info('Pure CTC mode')
+    elif args.mtlalpha == 0.0:
+        mtl_mode = 'att'
+        logging.info('Pure attention mode')
+    else:
+        mtl_mode = 'mtl'
```
Comment (kan-bayashi) [inline src/asr/asr_chainer.py]: It is also better to add logging.

File: src/nets/e2e_asr_attctc.py (hunk 1 above)
Comment (sw005320) [reply src/nets/e2e_asr_attctc.py]: This requires a lot of changes including several if-else, and I want to keep the reporter part simple. Also, it is not bad that all modes have the same format.

Comment (kan-bayashi) [reply src/nets/e2e_asr_attctc.py]: OK.

**Comment by sw005320** [comment]
> My rest concern is that acc.best is not available in pure ctc mode.
> Is it difficult to calculate accuracy using greedy ctc decoder?

This is not easy. Probably, one of the future work.

**Comment by shrutijpalaskar** [comment]
@sw005320 what would be a good way to monitor pure CTC training? As the accuracy is not calculated, how should we decide learning rate decay and convergence etc.?

Also, to confirm, for pure CTC decoding, the only change necessary is that ctc_weight should be set to 1.0, correct?

**Comment by sw005320** [comment]
> @sw005320 what would be a good way to monitor pure CTC training? As the accuracy is not calculated, how should we decide learning rate decay and convergence etc.?

So far, we're just using the conditional likelihood (CTC loss) for monitoring the CTC training and learning rate decay.

> Also, to confirm, for pure CTC decoding, the only change necessary is that ctc_weight should be set to 1.0, correct?

You need to use both `--ctc_weight 1.0` and `--recog_model loss.best`
